eventio/simtel/simtelfile.py

```diff
@@ -0,0 +1,394 @@
+        ]
+
+        self.shower = None
+        self.tel_moni = {}  # tel_id: SimTelTelMoni
+        self.lascal = {}  # tel_id: SimTelLasCal
+
+        self.cam_settings = {}
+        for telescope_description in self.telescope_descriptions:
+            cam_setting = telescope_description[0]
+            self.cam_settings[cam_setting['telescope_id']] = cam_setting
+
+        self.ref_pulse = {}
+        self.time_slices_per_telescope = {}
+        for telescope_description in self.telescope_descriptions:
+            pixel_setting = telescope_description[2]
+            telescope_id = pixel_setting['telescope_id']
+            self.time_slices_per_telescope[telescope_id] = pixel_setting['time_slice']
+            self.ref_pulse[telescope_id] = {
+                'step': pixel_setting['ref_step'],
+                'shape': pixel_setting['refshape']
+            }
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        while True:
+            try:
+                shower, event = self.fetch_next_event()
+                if 'event' in event:
+                    return shower, event
+            except NoTrackEvents:
+                logging.warning('skipping event: no TrackEvents')
+    def fetch_next_event(self):
+        try:
+            event = self.next_mc_event()
+            return self.shower, event
+        except WrongType:
+            try:
+                self.shower = self.file_.next_assert(SimTelMCShower).parse_data_field()
+            except WrongType:
+                raise StopIteration
+
+            event = self.next_mc_event()
+            return self.shower, event
+
+    def next_mc_event(self):
+        result = {}
+
+        # There is for sure exactly one of these
+        result['mc_event'] = self.file_.next_assert(SimTelMCEvent).parse_data_field()
+        result['corsika_tel_data'] = self.file_.next_type_or_none(
+            CORSIKATelescopeData)
+
+        self.update_moni_lascal()
+        try:
+            result['pe_sum'] = self.file_.next_assert(SimTelMCPeSum).parse_data_field()
+        except WrongType:
+            return result
+
+        event_ = self.file_.next_type_or_none(SimTelEvent)
+        if event_ is not None:
+            result['event'] = parse_event(event_)
+
+        return result
+
+    def update_moni_lascal(self):
+        try:
+            for tel_id in range(self.n_telescopes):
+                moni = self.file_.next_assert(
+                    SimTelTelMoni
+                ).parse_data_field()
+                self.tel_moni[moni['telescope_id']] = moni
+
+                lascal = self.file_.next_assert(
+                    SimTelLasCal
+                ).parse_data_field()
+                self.lascal[lascal['telescope_id']] = lascal
+
+        except WrongType:
+            # this is normal .. not every event has updates here
+            pass
+
+
+
+class EventIOFileWithNextAssert(EventIOFile, WithNextAssert):
```

Comment:
same here, does this new functionality have a large overhead? Or could we just put this in `base.py`?